### PR TITLE
Make gen_batch_initial_conditions more flexible

### PR DESCRIPTION
**Pull request by jduerholt**
## Motivation

This PR adds the feature regarding additional flexibility of `gen_batch_initial_conditions` as discussed in issue https://github.com/pytorch/botorch/issues/1776.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/pytorch/botorch/blob/main/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

Unit tests.

### CODE REVIEW

**Comment by jduerholt** [comment]
Hmm, is the error in the tutorial pipeline really caused by this PR, no or?

One other thing: 

I just saw that `optimize_acqf_mixed` do not have the `ic_generator` and `ic_gen_kwargs` arguments. From my perspective, they should be added also there, or?

**Comment by Balandat** [comment]
> Hmm, is the error in the tutorial pipeline really caused by this PR, no or?

No you're all good, this was an issue due to some pandas upgrade, this was fixed in #1777.

File: botorch/optim/initializers.py [hunk 1]
```diff
@@ -274,6 +275,9 @@ def gen_batch_initial_conditions(
         equality constraints: A list of tuples (indices, coefficients, rhs),
             with each tuple encoding an inequality constraint of the form
             `\sum_i (X[indices[i]] * coefficients[i]) = rhs`.
+        generator: Callable for generating samples that are then further
+            processed. It receives `n`, `q` and `seed` as arguments and  returns
```
Comment (Balandat) [inline botorch/optim/initializers.py:279]: should we make `seed` optional here?

File: botorch/optim/initializers.py [hunk 2]
```diff
@@ -297,6 +301,11 @@ def gen_batch_initial_conditions(
             "Option 'sample_around_best' is not supported when equality"
             "constraints are present."
         )
+    if sample_around_best and generator:
+        raise UnsupportedError(
+            "Option 'sample_around_best' is not supported when custom"
+            "generator should be used."
```
Comment (Balandat) [inline botorch/optim/initializers.py:307]:
```suggestion
            "Option 'sample_around_best' is not supported when custom "
            "generator is be used."
```

File: botorch/optim/initializers.py [hunk 3]
```diff
@@ -327,7 +336,9 @@ def gen_batch_initial_conditions(
     while factor < max_factor:
         with warnings.catch_warnings(record=True) as ws:
             n = raw_samples * factor
-            if inequality_constraints is None and equality_constraints is None:
+            if generator:
```
Comment (Balandat) [inline botorch/optim/initializers.py:339]:
```suggestion
            if generator is not None:
```

File: test/optim/test_initializers.py [hunk 4]
```diff
@@ -595,6 +595,80 @@ def test_gen_batch_initial_conditions_interpoint_constraints(self):
...
+            gen_batch_initial_conditions(
+                MockAcquisitionFunction(),
+                bounds=torch.tensor([[0, 0], [1, 1]], **tkwargs),
+                q=1,
+                num_restarts=1,
+                raw_samples=1,
+                generator=generator,
+                options={"sample_around_best": True},
+            )
+            self.assertTrue(
```
Comment (Balandat) [inline test/optim/test_initializers.py:667]: could instead use `assertRaisesRegex`, which makes this easier / cleaner.

**Review by Balandat — APPROVED** [review]
a few nits and and there, but otherwise this looks great. Thanks!

File: test/optim/test_initializers.py (hunk 4 above)
Comment (jduerholt) [reply test/optim/test_initializers.py]: Done.

File: botorch/optim/initializers.py (hunk 1 above)
Comment (jduerholt) [reply botorch/optim/initializers.py]: I would keep it mandatory, as a seed is always used in `gen_batch_initial_conditions` and the seed number is increased in every iteration.

**Comment by Balandat** [comment]
Thanks @esantorella! Should be an easy fix; once that's in this should be good to land!

**Comment by jduerholt** [comment]
Thanks @esantorella for fixing it. Is there anything to do for me or is now everything fixed?

**Comment by esantorella** [comment]
> Thanks @esantorella for fixing it. Is there anything to do for me or is now everything fixed?

Looks to me like everything is fixed now! Nothing more should be needed on your end.
